crates/roc_host/src/json_web_token.rs

```diff
@@ -0,0 +1,170 @@
+use hmac::{Hmac, Mac};
+use jwt::algorithm::AlgorithmType;
+use roc_std::{RocList, RocResult, RocStr};
+use sha2::{Sha256, Sha384, Sha512};
+use std::collections::BTreeMap;
+
+#[derive(Clone, Default, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
+#[repr(C)]
+pub struct JWTFromRoc {
+    secret: RocStr,
+    token: RocStr,
+    algo: u8,
+}
+
+impl JWTFromRoc {
+    fn algorithm(&self) -> AlgorithmType {
+        match self.algo {
+            1 => AlgorithmType::Hs256,
+            2 => AlgorithmType::Hs384,
+            3 => AlgorithmType::Hs512,
+            4 => AlgorithmType::Rs256,
+            5 => AlgorithmType::Rs384,
+            6 => AlgorithmType::Rs512,
+            7 => AlgorithmType::Es256,
+            8 => AlgorithmType::Es384,
+            9 => AlgorithmType::Es512,
+            10 => AlgorithmType::Ps256,
+            11 => AlgorithmType::Ps384,
+            12 => AlgorithmType::Ps512,
+            13 => AlgorithmType::None,
+            _ => panic!("invalid algorithm from roc"),
+        }
+    }
+}
+
+impl roc_std::RocRefcounted for JWTFromRoc {
+    fn inc(&mut self) {
+        self.secret.inc();
+        self.token.inc();
+    }
+    fn dec(&mut self) {
+        self.secret.dec();
+        self.token.dec();
+    }
+    fn is_refcounted() -> bool {
+        true
+    }
+}
+
+#[derive(Clone, Default, Debug, PartialEq, PartialOrd, Eq, Ord, Hash)]
+#[repr(C)]
+pub struct JWTToRoc {
+    name: RocStr,
+    value: RocStr,
+}
+
+impl roc_std::RocRefcounted for JWTToRoc {
+    fn inc(&mut self) {
+        self.name.inc();
+        self.value.inc();
+    }
+    fn dec(&mut self) {
+        self.name.dec();
+        self.value.dec();
+    }
+    fn is_refcounted() -> bool {
+        true
+    }
+}
+
+fn jwt_claims_to_roc(claims: BTreeMap<String, serde_json::Value>) -> RocList<JWTToRoc> {
+    let mut roc_claims = RocList::with_capacity(claims.len());
+
+    for (name, json_value) in claims {
+        roc_claims.push(JWTToRoc {
+            name: name.as_str().into(),
+            value: format!("{}", json_value).as_str().into(),
```

Comment:
Why does this require format! for a single value?

Reply:
It's a serve_json::Value... and I tried to do it few other ways and couldn't make it work. Doing a simple `.as_str().unwrap()` sometimes it would try to unwrap a None and crash. I definitely don't love this implementation, because it formats strings with surrounding double quotes, so `"mystring"` instead of `mystring` etc.

Reply:
Interesting.